File: src/syncer.js
/**
 * Background fetching, the same shape as trainer.js: `requestSync` returns at
 * once and the fetching runs in a worker thread on its own SQLite connection,
 * so a request never waits on a few hundred sequential HTTP calls.
 *
 * Unlike training, requests are *not* coalesced. A sync carries parameters
 * (a window, a range, a points floor), so a second request while one runs
 * cannot be folded into it — it is refused as `busy` and the caller can retry.
 * Only one run at a time either way: two would fight over the same days.
 */
import { Worker } from 'node:worker_threads';
import { dbPath } from './db.js';

const state = {
  running: false,
  startedAt: null,
  options: null,      // what the current (or last) run was asked to do
  progress: null,     // most recent { day, count, failed? }
  last: null,         // result of the most recent completed run
  lastError: null,
  runs: 0,
};

const waiters = [];

/** Start a sync unless one is running. Returns the status either way. */
export function requestSync(options = {}) {
  if (state.running) return { status: 'busy', ...syncStatus() };

  state.running = true;
  state.startedAt = Date.now();
  state.options = options;
  state.progress = null;
  const worker = new Worker(new URL('./sync-worker.js', import.meta.url), {
    workerData: { dbPath: dbPath(), options },
  });
  // The worker streams progress and reports exactly once at the end:
  // 'message' with a result normally, 'error'/'exit' only if it died first.
  let done = false;
  const finish = (msg) => {
    if (done) return;
    done = true;
    state.running = false;
    state.startedAt = null;
    state.runs++;
    if (msg.ok) {
      state.last = { ...msg.result, finishedAt: Math.floor(Date.now() / 1000) };
      state.lastError = null;
    } else {
      state.lastError = msg.error;
      console.error('sync failed:', msg.error);
    }
    while (waiters.length) waiters.shift()();
  };
  worker.on('message', (msg) => {
    if (msg.progress) state.progress = msg.progress;
    else finish(msg);
  });
  worker.once('error', (err) => finish({ ok: false, error: err.message ?? String(err) }));
  worker.once('exit', (code) => finish({ ok: false, error: `sync worker exited with code ${code}` }));

  return { status: 'started', ...syncStatus() };
}

export function syncStatus() {
  return {
    running: state.running,
    startedAt: state.startedAt,
    options: state.options,
    progress: state.progress,
    runs: state.runs,
    last: state.last,
    lastError: state.lastError,
  };
}

/** Resolve once no sync is running (tests, CLI). */
export function syncIdle() {
  if (!state.running) return Promise.resolve();
  return new Promise((resolve) => waiters.push(resolve));
}

File: src/syncer.rs
//! Background fetching, the same shape as trainer.rs: `request` returns at
//! once and the fetching runs on its own thread with its own database connection,
//! so a request never waits on a few hundred sequential HTTP calls.
//!
//! Unlike training, requests are *not* coalesced. A sync carries parameters
//! (a window, a range, a points floor), so a second request while one runs
//! cannot be folded into it — it is refused as `busy` and the caller can retry.
//! Only one run at a time either way: two would fight over the same days.

use std::panic::AssertUnwindSafe;
use std::sync::{Arc, Condvar, Mutex};

use serde_json::{json, Value};

use crate::dates::{now_millis, now_seconds};
use crate::db::open_db;
use crate::hn::Algolia;
use crate::http_client::HttpFetcher;
use crate::service::{sync, ModelCache, SyncRequest};
use crate::trainer::panic_message;

#[derive(Default)]
struct SyncState {
    running: bool,
    started_at: Option<i64>, // unix millis
    options: Option<Value>,  // what the current (or last) run was asked to do
    progress: Option<Value>, // most recent { day, count, failed? }
    runs: u64,
    last: Option<Value>, // result of the most recent completed run
    last_error: Option<String>,
}

pub struct Syncer {
    state: Mutex<SyncState>,
    idle: Condvar,
    db_url: String,
    cache: Arc<ModelCache>,
}

impl Syncer {
    /// The state, with poison recovery — same reasoning as `Trainer::state`.
    fn state(&self) -> std::sync::MutexGuard<'_, SyncState> {
        self.state
            .lock()
            .unwrap_or_else(|poisoned| poisoned.into_inner())
    }

    pub fn new(db_url: String, cache: Arc<ModelCache>) -> Arc<Syncer> {
        Arc::new(Syncer {
            state: Mutex::new(SyncState::default()),
            idle: Condvar::new(),
            db_url,
            cache,
        })
    }

    /// Start a sync unless one is running. Returns the status either way.
    pub fn request(self: &Arc<Self>, req: SyncRequest) -> Value {
        let mut state = self.state();
        if state.running {
            let mut out = status_json(&state);
            out["status"] = json!("busy");
            return out;
        }
        state.running = true;
        state.started_at = Some(now_millis());
        state.options = Some(req.to_json());
        state.progress = None;
        let syncer = Arc::clone(self);
        std::thread::spawn(move || syncer.run(req));
        let mut out = status_json(&state);
        out["status"] = json!("started");
        out
    }

    fn run(self: Arc<Self>, req: SyncRequest) {
        let started = now_millis();
        let result = std::panic::catch_unwind(AssertUnwindSafe(|| {
            let conn = open_db(&self.db_url);
            let fetcher = HttpFetcher::default();
            let source = Algolia { fetch: &fetcher };
            let progress_sink = &self;
            sync(&conn, &self.cache, &req, &source, &mut |p| {
                progress_sink.state().progress =
                    Some(serde_json::to_value(p).expect("progress json"));
            })
        }));
        let mut state = self.state();
        state.runs += 1;
        match result {
            Ok(Ok(outcome)) => {
                let mut value = serde_json::to_value(&outcome).expect("sync json");
                value["ms"] = json!(now_millis() - started);
                value["finishedAt"] = json!(now_seconds());
                state.last = Some(value);
                state.last_error = None;
            }
            Ok(Err(message)) => {
                eprintln!("sync failed: {message}");
                state.last_error = Some(message);
            }
            Err(payload) => {
                let message = panic_message(payload);
                eprintln!("sync failed: {message}");
                state.last_error = Some(message);
            }
        }
        state.running = false;
        state.started_at = None;
        self.idle.notify_all();
    }

    pub fn status(&self) -> Value {
        status_json(&self.state())
    }

    /// Block until no sync is running (tests, CLI).
    pub fn wait_idle(&self) {
        let mut state = self.state();
        while state.running {
            state = self
                .idle
                .wait(state)
                .unwrap_or_else(|poisoned| poisoned.into_inner());
        }
    }
}

fn status_json(state: &SyncState) -> Value {
    json!({
        "running": state.running,
        "startedAt": state.started_at,
        "options": state.options,
        "progress": state.progress,
        "runs": state.runs,
        "last": state.last,
        "lastError": state.last_error,
    })
}
